File: python/indx_list.py
from __future__ import annotations

#  Originally, this code contained a single class with encapsulated fields,
#  but after profiling it turned out that calling ANY function from inside of a class
#  spends time comparable to one call to rand_part() function  of QuickSort algo.
#  As a result, I switched to Plain C-like structure (like in indx_list.h)

class indx_item:
    def __init__(self, indx: int):
        self.indx = indx
        self.next = None
        self.prev = None


def indx_add_next(self, indx: int) -> indx_item:
    res = indx_item(indx)
    res.next = self.next
    res.prev = self
    self.next = res
    return res


def indx_del_item(self):
    if self.prev is not None and self.next is not None:
        self.prev.next = self.next
        self.next.prev = self.prev
    elif self.next is not None:
        assert self.prev is None
        self.next.prev = None
    elif self.prev is not None:
        assert self.next is None
        self.prev.next = None
    else:
        assert self.prev is None
        assert self.next is None

        # Can't "free(self)" in python :)


def indx_print(el):
    while el.prev is not None: el = el.prev
    while el is not None:
        print(f"{el.indx}, ", end="")
        el = el.next
    print("")

File: c/indx_list.c
//
// Created by Irene on 22/10/2023.
//

#include "indx_list.h"

#include <assert.h>
#include <stdlib.h>
#include <stdio.h>

struct indx_item* indx_create_initial(size_t indx) {
    struct indx_item* res = malloc(sizeof(struct indx_item));
    res->indx = indx;
    res->next = NULL;
    res->prev = NULL;
    return res;
}

struct indx_item* indx_create_next(struct indx_item *current, size_t indx) {
    struct indx_item* res = malloc(sizeof(struct indx_item));
    res->indx = indx;
    res->next = current->next;
    res->prev = current;
    current->next = res;
    return res;
}

void indx_destroy_item(struct indx_item *current) {
    if (current->prev != NULL && current->next != NULL) {
        current->prev->next = current->next;
        current->next->prev = current->prev;
    } else if (current->next != NULL) {
        assert(current->prev == NULL);
        current->next->prev = NULL;
    } else if (current->prev != NULL) {
        assert(current->next == NULL);
        current->prev->next = NULL;
    } else {
        assert(current->prev == NULL);
        assert(current->next == NULL);
    }

    free(current);
}

void indx_destroy_all(struct indx_item *el) {
    while (el->prev != NULL) { el = el->prev; }
    while (el != NULL) {
        struct indx_item* el_next = el->next;
        free(el);
        el = el_next;
    }
}

void indx_print(struct indx_item *el) {
    while (el->prev != NULL) { el = el->prev; }
    while (el != NULL) {
        printf("%zu, ", el->indx);
        el = el->next;
    }
    printf("\n");
}

